Source/Python/WarpXWrappers.cpp

```diff
@@ -652,6 +652,37 @@ namespace
         particle_buffers.clearParticles();
     }
 
+    void warpx_depositChargeDensity (const char* char_species_name, int lev) {
+        // this function is used to deposit a given species' charge density
+        // in the rho_fp multifab which can then be accessed from python via
+        // pywarpx.fields.RhoFPWrapper()
+        WarpX& warpx = WarpX::GetInstance();
+        const auto & mypc = warpx.GetPartContainer();
+        const std::string species_name(char_species_name);
+        auto & myspc = mypc.GetParticleContainerFromName(species_name);
+        auto * rho_fp = warpx.get_pointer_rho_fp(lev);
+
+        if (rho_fp == nullptr) {
+            warpx.RecordWarning(
```

Comment:
Is there some way that when this happens, an exception is raised at the Python level? Perhaps this routine could return an error code? Alternatively, could `rho_fp` be allocated here? This would allow this feature to be used in cases when `rho_fp` would otherwise not be allocated.

Reply:
In pybind11, `throw`-in an exception from C++ translates into a Python exception. Not sure what happens currently if we throw in our bindings, I guess we just abort then?

Is the case here severe? In that case, at least `WarnPriority::low` might better be `WarnPriority::high`?

Reply:
Currently if an exception is raised in a Python callback the code execution is not terminated, so raising an error at the Python level would mostly be equivalent to the current `warpx.RecordWarning` approach.
If I raise an exception from C++, that does terminate the code. But I don't think we should do that here since this functionality is meant to be used for diagnostics so I think that printing the warning saying `rho_fp` is not allocated is sufficient.

I agree allocating `rho_fp` in this call, if needed, would be nice. The immediate difficulty I see is that `WarpX::AllocLevelData` takes a `BoxArray` and `DistributionMapping` as arguments that is used when allocating `rho_fp`. I'm not sure how to retrieve those objects in this call. Maybe we can merge this as is and if at some point someone runs into an issue where they would like to use this function but  `rho_fp` is not allocated, we can revisit this point?

Reply:
Merging now would be fine. I'll approve the PR.

Reply:
Thanks!